tgui/packages/tgui/interfaces/PersonalCrafting.tsx

```diff
@@ -1,311 +1,922 @@
-    const dm_categories_item = dm_categories[dm_category];
-
-    // This is a nested part of the tree containing subcategories
-    if ('has_subcats' in dm_categories_item) {
-      // Create a new group for these categories
-      const group: Group = {
-        name: dm_category,
-        categories: [],
-      };
-      groups.push(group);
-      for (const dm_subcategory of Object.keys(dm_categories_item)) {
-        // Skip even more nested subcats, they shouldn't exist
-        if (dm_subcategory === 'has_subcats') {
-          continue;
-        }
-        // Push category
-        group.categories.push({
-          name: dm_subcategory,
-          dm_category,
-          dm_subcategory,
-        });
-        // Push recipes
-        const dm_recipes = dm_categories_item[dm_subcategory];
-        for (const dm_recipe of dm_recipes) {
-          recipes.push({
-            ...dm_recipe,
-            dm_category,
-            dm_subcategory,
-            craftable: Boolean(craftability_by_ref[dm_recipe.ref]),
-          });
-        }
-      }
-      continue;
-    }
-
-    // This is a flat part of the tree where simple categories are located.
-    // We group these categories under the "General" group.
-    if (Array.isArray(dm_categories_item)) {
-      // Push category
-      generalGroup.categories.push({
-        name: dm_category,
-        dm_category,
-      });
-      // Push recipes
-      for (const dm_recipe of dm_categories_item) {
-        recipes.push({
-          ...dm_recipe,
-          dm_category,
-          craftable: Boolean(craftability_by_ref[dm_recipe.ref]),
-        });
-      }
-      continue;
-    }
-
-    throw new Error(
-      `Invalid entry in 'crafting_recipes', neither '{ has_subcats: 1 }' nor 'Recipe[]'`
-    );
-  }
-
-  return {
-    ...rawData,
-    dm_category: rawData.category,
-    dm_subcategory: rawData.subcategory,
-    recipes,
-    groups,
-  };
-};
-
-const isCategorySelected = (data: Data, item: Category | Recipe) =>
-  data.dm_category === item.dm_category &&
-  (!data.dm_subcategory || data.dm_subcategory === item.dm_subcategory);
-
 export const PersonalCrafting = (props, context) => {
-  const { act, data: rawData } = useBackend<RawData>(context);
-  const data = remapData(rawData);
-
-  const { busy, display_craftable_only, display_compact, recipes, groups } =
-    data;
-
-  const shownRecipes = flow([
+  const { act, data } = useBackend<Data>(context);
+  const {
+    mode,
+    busy,
+    display_compact,
+    display_craftable_only,
+    craftability,
+    diet,
+  } = data;
+  const [searchText, setSearchText] = useLocalState(context, 'searchText', '');
+  const [pages, setPages] = useLocalState(context, 'pages', 1);
+  const DEFAULT_CAT_CRAFTING = Object.keys(CATEGORY_ICONS_CRAFTING)[1];
+  const DEFAULT_CAT_COOKING = Object.keys(CATEGORY_ICONS_COOKING)[1];
+  const [activeCategory, setCategory] = useLocalState<string | undefined>(
```

Comment:
string always seems to have to be defined here, so why do string|undefined?

Reply:
It was used for some hacky way of switching between crafting and cooking, but no longer needed. I'll remove it.